R/paginate.R

```diff
@@ -0,0 +1,229 @@
+    req <- paginate_next_request(resp, req)
+    if (is.null(req)) {
+      page <- page - 1L
+      break
+    }
+
+    resp <- req_perform(req)
+
+    body_parsed <- resp_body_json(resp)
+    out[[page]] <- resp
+
+    cli::cli_progress_update()
+  }
+  cli::cli_progress_done()
+
+  # remove unused end of `out` in case the pagination loop exits before all
+  # `max_pages` is reached
+  if (page < n_pages) {
+    out <- out[seq2(1, page)]
+  }
+
+  out
+}
+
+#' @rdname paginate_perform
+paginate_next_request <- function(resp, req) {
+  check_response(resp)
+  check_request(req)
+
+  if (!req_policy_exists(req, "paginate")) {
+    cli::cli_abort(c(
+      "{.arg req} doesn't have a pagination policy",
+      i = "You can add pagination via `req_paginate()`."
+    ))
+  }
+
+  next_request <- req$policies$paginate$next_request
+  next_request(resp = resp, req = req)
+}
+
+#' @rdname req_paginate
+#' @export
+req_paginate_next_url <- function(req,
+                                  next_url,
+                                  n_pages = NULL) {
+  check_function2(next_url, args = "resp")
+
+  next_request <- function(req, resp) {
+    next_url <- next_url(resp)
+
+    if (is.null(next_url)) {
+      return(NULL)
+    }
+
+    req_url(req, next_url)
+  }
+
+  req_paginate(
+    req,
+    next_request,
+    n_pages = n_pages
+  )
+}
+
+#' @rdname req_paginate
+#' @export
+req_paginate_offset <- function(req,
+                                offset,
+                                page_size,
+                                n_pages = NULL) {
+  check_number_whole(page_size)
+  check_function2(offset, args = c("req", "offset"))
+
+  cur_offset <- 0L
+  env <- current_env()
+  next_request <- function(resp, req) {
+    cur_offset <- get("cur_offset", envir = env)
+    new_offset <- cur_offset + page_size
+    assign("cur_offset", new_offset, envir = env)
+
+    offset(req, new_offset)
+  }
+
+  req_paginate(
+    req,
+    next_request,
+    n_pages
+  )
+}
+
+#' @rdname req_paginate
+#' @export
+req_paginate_next_token <- function(req,
+                                    set_token,
+                                    next_token,
+                                    n_pages = NULL) {
+  check_function2(set_token, args = c("req", "token"))
+  check_function2(next_token, args = "resp")
+
+  next_request <- function(req, resp) {
+    next_token <- next_token(resp)
+
+    if (is.null(next_token)) {
+      return(NULL)
+    }
+
+    set_token(req, next_token)
+  }
+
+  req_paginate(
+    req,
+    next_request,
+    n_pages = n_pages
+  )
+}
+
+#' Calculate the number of pages
+#'
+#' @param page_size An integer.
+#' @param total A function that extracts the total number of elements from the [response].
+#'
+#' @return A function that can be passed to the `n_pages` argument of [req_paginate()].
+#' @export
+#'
+#' @examples
+#' calculate_n_pages(page_size = 100, total = 250)
+calculate_n_pages <- function(page_size, total) {
```

Comment:
Before this took a response object and `total` was a function that takes the response as an argument.
In this form it is hardly worth exporting. On the other hand when doing pagination one usually gets either the total number of elements (and needs to know the page size) or the total number of pages.

Reply:
I'd say not to export this for now.